src/bp/core/repositories/conversations.ts

```diff
@@ -0,0 +1,190 @@
+    const numberOfDeletedRows = await this.query()
+      .where(endpoint)
+      .del()
+
+    this.invalidateConvCache(undefined)
+
+    return numberOfDeletedRows
+  }
+
+  public async create(endpoint: sdk.UserEndpoint): Promise<sdk.Conversation> {
+    let row = {
+      userId: endpoint.userId,
+      botId: endpoint.botId,
+      createdOn: new Date()
+    }
+
+    const [id] = await this.query()
+      .insert(this.serialize(row))
+      .returning('id')
+
+    const conversation = {
+      id,
+      ...row
+    }
+    this.cache.set(id, conversation)
+
+    return conversation
+  }
+
+  public async getMostRecent(endpoint: sdk.UserEndpoint): Promise<sdk.Conversation | undefined> {
+    let query = this.queryRecents(endpoint)
+    query = query.limit(1)
+
+    return this.deserialize((await query)[0])
+  }
+
+  public async getById(conversationId: number): Promise<sdk.Conversation | undefined> {
+    const cached = this.cache.get(conversationId)
+    if (cached) {
+      return cached
+    }
+
+    const rows = await this.query()
+      .select('*')
+      .where({ id: conversationId })
+
+    const conversation = this.deserialize(rows[0])
+    if (conversation) {
+      this.cache.set(conversationId, conversation)
+    }
+
+    return conversation
+  }
+
+  public async delete(conversationId: number): Promise<boolean> {
+    const numberOfDeletedRows = await this.query()
+      .where({ id: conversationId })
+      .del()
+
+    this.invalidateConvCache(conversationId)
+
+    return numberOfDeletedRows > 0
+  }
+
+  private queryRecents(endpoint: sdk.UserEndpoint) {
+    return this.query()
+      .select(
+        'conversations.id',
+        'conversations.userId',
+        'conversations.botId',
+        'conversations.createdOn',
+        'messages.id as messageId',
+        'messages.eventId',
+        'messages.incomingEventId',
+        'messages.from',
+        'messages.payload',
+        'messages.sentOn'
+      )
+      .leftJoin('messages', 'messages.conversationId', 'conversations.id')
+      .where({
+        userId: endpoint.userId,
+        botId: endpoint.botId,
+        sentOn: this.database.knex
+          .max('sentOn')
+          .from('messages')
+          .where('messages.conversationId', this.database.knex.ref('conversations.id'))
+      })
+      .groupBy('conversations.id', 'messages.id')
+      .orderBy('sentOn', 'desc')
+  }
+
+  public query() {
+    return this.database.knex(this.TABLE_NAME)
+  }
+
+  public serialize(conversation: Partial<sdk.Conversation>) {
+    const { userId, botId, createdOn } = conversation
+    return {
+      userId,
+      botId,
+      createdOn: createdOn?.toISOString()
+    }
+  }
+
+  public deserialize(conversation: any): sdk.Conversation | undefined {
+    if (!conversation) {
+      return undefined
+    }
+
+    const { id, userId, botId, createdOn } = conversation
+    return {
+      id,
+      userId,
+      botId,
+      createdOn: new Date(createdOn)
+    }
+  }
```

Comment:
I added a get and set function for dates. new Date() and toIsoString() seem to work perfectly fine for both sqlite and postgres. I don't like using knex.date.now() because I need to return the date to the function caller. So I would end up doing knex.date.now() and then knex.date.get().